### PR TITLE
HDDS-3705. [OFS] Implement getTrashRoots for trash cleanup

### PR DESCRIPTION
**Pull request by smengcl**
## What changes were proposed in this pull request?

Implement getTrashRoots for trash cleanup

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3705

## How was this patch tested?

Will add a new test.

### CODE REVIEW

**Comment by smengcl** [comment]
I'm rebasing the branch since the last update is a while ago. Also there were some flaky ozone-recon robot test in the [first run](https://github.com/apache/hadoop-ozone/runs/782209210).

**Comment by smengcl** [comment]
This PR is ready for review.

**Comment by smengcl** [comment]
The new test contaminated the results of other tests in the same test class. Fixing this.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java [hunk 1]
```diff
@@ -871,4 +876,103 @@ public void testFailToDeleteRoot() throws IOException {
...
+    Path user1Trash1 = new Path(trashRoot1, username);
+    // When user trash dir isn't been created
+    Assert.assertEquals(0, fs.getTrashRoots(false).size());
+    Assert.assertEquals(0, fs.getTrashRoots(true).size());
+    // Let's create our first user1 (current user) trash dir.
+    fs.mkdirs(user1Trash1);
+    // Results should be getTrashRoots(false)=1, gTR(true)=1
+    Collection<FileStatus> res = fs.getTrashRoots(false);
+    Assert.assertEquals(1, res.size());
+    res.forEach(e -> Assert.assertEquals(
```
Comment (xiaoyuyao) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java:900]: NIT: forEach is unnecessary as we have assert the size is 1.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java [hunk 2]
```diff
@@ -871,4 +876,103 @@ public void testFailToDeleteRoot() throws IOException {
...
+    res = fs.getTrashRoots(true);
+    Assert.assertEquals(1, res.size());
+    res.forEach(e -> Assert.assertEquals(
+        user1Trash1.toString(), e.getPath().toUri().getPath()));
+
+    // Create one more trash for user2 in the same bucket
+    Path user2Trash1 = new Path(trashRoot1, "testuser2");
+    fs.mkdirs(user2Trash1);
+    // Results should be getTrashRoots(false)=1, gTR(true)=2
+    Assert.assertEquals(1, fs.getTrashRoots(false).size());
```
Comment (xiaoyuyao) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java:911]: Can we assert the trash roots is for user one with invoke getTrashRoots(false)?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java [hunk 3]
```diff
@@ -871,4 +876,103 @@ public void testFailToDeleteRoot() throws IOException {
...
+    Path bucketPath2 = new Path(volumePath, bucketName2);
+    Path trashRoot2 = new Path(bucketPath2, TRASH_PREFIX);
+    Path user1Trash2 = new Path(trashRoot2, username);
+    // Create a file at the trash location, it shouldn't be recognized as trash
+    try (FSDataOutputStream out1 = fs.create(user1Trash2)) {
+      out1.write(123);
+    }
+    // Results should still be getTrashRoots(false)=1, gTR(true)=2
+    Assert.assertEquals(1, fs.getTrashRoots(false).size());
+    res.forEach(e -> Assert.assertEquals(
```
Comment (xiaoyuyao) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java:926]: I see you have the assertion here for user1Trash1. So you can ignore the previous comment.

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java [hunk 4]
```diff
@@ -516,6 +520,62 @@ public FileStatusAdapter getFileStatus(String path, URI uri,
...
+            Path userTrash = new Path(trashRoot, username);
+            if (fs.exists(userTrash) &&
+                fs.getFileStatus(userTrash).isDirectory()) {
+              ret.add(fs.getFileStatus(userTrash));
+            }
+          }
+        }
+      }
+    } catch (IOException ex) {
+      throw new RuntimeException(ex);
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java:574]: I don't think we should throw RuntimeException here. We can log a warning and return an empty collection instead.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java (hunk 1 above)
Comment (smengcl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java]: done d997ac0fb09b0145824ae8196e6b9bb47956c086

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java (hunk 2 above)
Comment (smengcl) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/fs/ozone/TestRootedOzoneFileSystem.java]: sure thing. done in ebf104b3d6b0c6531ce2bec33fe00c996fcb16cf

File: hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java (hunk 4 above)
Comment (smengcl) [reply hadoop-ozone/ozonefs-common/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java]: done 1940f99f524ef589b2c5a1613031f3cb7b1066b4

**Comment by xiaoyuyao** [comment]
LGTM, +1 pending CI.

**Comment by smengcl** [comment]
Thanks @xiaoyuyao  for the review. Will merge shortly.

**Pull request merged**